File: libmedia/media/platform/android/video.c

#include "my_handle.h"
#include "mbuf.h"
#include "media_buffer.h"
#include "pts.h"
#include "lock.h"
#include "myjni.h"
#include "pts.h"
#include "video.h"
#include "android.h"
#include "thread_clean.h"

typedef struct {
    lock_t lck;
    intptr_t run;
    video_push_t fptr_push;
    void* uptr_push;

    enum camera_position_t campos;
    intptr_t angle;
    intptr_t mirror;
    my_handle* self;

    video_format* fmt;
    intptr_t width, height;
    uint32_t stride[3];

    uint32_t seq;
    media_iden_t id;
    java_code jcode;
} video_capture_t;
extern jclass media_clszz;

static struct my_buffer* java_buffer_clone(struct my_buffer* mbuf)
{
    java_buffer_t* jbuf = (java_buffer_t *) (mbuf->ptr[1] + mbuf->length);
    mbuf = jbuf->mop->clone(mbuf);
    if (mbuf != NULL) {
        __sync_add_and_fetch(&jbuf->refs, 1);
    }
    return mbuf;
}

static void java_buffer_release(struct my_buffer* mbuf)
{
    java_buffer_t* jbuf = (java_buffer_t *) (mbuf->ptr[1] + mbuf->length);
    intptr_t n = __sync_sub_and_fetch(&jbuf->refs, 1);
    if (n == 0) {
        java_code stack_java;
        java_code* code = thread_clean_push(0, (free_t) jni_put, sizeof(java_code));
        if (code == NULL) {
            jcode_get_env(&stack_java);
            code = &stack_java;
        }

        JNIEnv* env = jni_get(code, NULL);
        my_assert(env != NULL);
        (*env)->ReleaseByteArrayElements(env, jbuf->pixel, jbuf->mem, 0);
        (*env)->DeleteGlobalRef(env, jbuf->pixel);
        if (code == &stack_java) {
            jni_put(code);
        }
    }
    mbuf->mop = jbuf->mop;
    mbuf->mop->free(mbuf);
}

static struct mbuf_operations java_buffer_mop = {
    .clone = java_buffer_clone,
    .free = java_buffer_release
};

jint jni_pixel_push(JNIEnv* env, jobject obj, jlong something, jbyteArray pixel)
{
    my_handle* handle = (my_handle *) (intptr_t) something;
    video_capture_t* capt = (video_capture_t *) handle_get(handle);
    if (capt == NULL) {
        return -1;
    }

    if (__builtin_expect(capt->angle == -1, 0)) {
        handle_put(handle);
        return -1;
    }

    if (capt->width == 0 || capt->height == 0) {
        handle_put(handle);
        return 0;
    }

    struct my_buffer* mbuf = NULL;
    uintptr_t bytes = 0;
    jint length = (*env)->GetArrayLength(env, pixel);
    if (reference_format(capt->fmt)) {
        bytes = sizeof(media_buffer) + sizeof(java_buffer_t);
    } else {
        bytes = sizeof(media_buffer) + length;
    }

    mbuf = mbuf_alloc_2(bytes);
    if (mbuf == NULL) {
        handle_put(handle);
        return 0;
    }
    mbuf->length = bytes - sizeof(media_buffer);
    mbuf->ptr[1] = mbuf->ptr[0] + sizeof(media_buffer);

    media_buffer* media = (media_buffer *) mbuf->ptr[0];
    memset(&media->vp, 0, sizeof(media->vp));
    media->frametype = 0;
    media->angle = capt->angle | capt->mirror;
    media->fragment[0] = 0;
    media->fragment[1] = 1;
    media->pptr_cc = &(capt->fmt->cc);
    media->iden = capt->id;
    media->seq = capt->seq++;
    fill_pts(NULL, mbuf);

    media->vp[0].stride = capt->stride[0];
    media->vp[0].height = capt->fmt->pixel->size->height;
    if (reference_format(capt->fmt)) {
        mbuf->length -= sizeof(java_buffer_t);

        java_buffer_t* jbuf = (java_buffer_t *) (mbuf->ptr[1] + mbuf->length);
        jbuf->mem = (*env)->GetByteArrayElements(env, pixel, NULL);
        jbuf->refs = 1;
        jbuf->pixel = (jbyteArray) (*env)->NewGlobalRef(env, pixel);
        jbuf->mop = mbuf->mop;
        mbuf->mop = &java_buffer_mop;
        media->vp[0].ptr = (char *) jbuf->mem;
    } else {
        (*env)->GetByteArrayRegion(env, pixel, 0, length, (jbyte *) mbuf->ptr[1]);
        media->vp[0].ptr = (char *) mbuf->ptr[1];
    }

    for (uint32_t i = 1; i < capt->fmt->pixel->panels; ++i) {
        media->vp[i].stride = capt->stride[i];
        media->vp[i].height = panel_height(i, capt->fmt->pixel->size->height, capt->fmt->pixel->csp);
        intptr_t h = panel_height(i - 1, capt->height, capt->fmt->pixel->csp);
        media->vp[i].ptr = media->vp[i - 1].ptr + (media->vp[i - 1].stride * h);
    }

    capt->fptr_push(capt->uptr_push, mbuf);
    handle_put(handle);
    return 0;
}

static intptr_t video_begin_capt(my_handle* handle, video_capture_t* capt)
{
    java_code* jcode = &capt->jcode;
    JNIEnv* env = jni_get(jcode, media_clszz);
    if (env == NULL) {
        return -1;
    }

    video_format* fmt = capt->fmt;
    jlong something = (jlong) (intptr_t) handle;
    jintArray result = (jintArray) (*env)->CallStaticObjectMethod(env, jcode->clazz, jcode->id, (jint) capt->campos,
                                                                  fmt->pixel->size->width, fmt->pixel->size->height, fmt->pixel->csp,
                                                                  fmt->caparam->fps, something);
    if (result == NULL) {
        jni_put(jcode);
        return -1;
    }

    jint nr = (*env)->GetArrayLength(env, result);
    jint* buf = (*env)->GetIntArrayElements(env, result, NULL);

    my_assert(nr > 2);
    intptr_t i = 0;
    capt->stride[0] = capt->stride[1] = capt->stride[2] = 0;
    switch (nr) {
        case 5: capt->stride[i] = buf[i]; i++;
        case 4: capt->stride[i] = buf[i]; i++;
        case 3: capt->stride[i] = buf[i]; i++;
    }
    capt->width = buf[i++];
    capt->height = buf[i];
    logmsg("%d, %d selected\n", (int) capt->width, (int) capt->height);

    (*env)->ReleaseIntArrayElements(env, result, buf, 0);
    jni_put(jcode);
    return 0;
}

static void capt_cleanup(void* addr)
{
    video_capture_t* capt = (video_capture_t *) addr;
    handle_release(capt->self);
    capt->self = NULL;

    capt->width = capt->height = 0;
    if (capt->fptr_push != NULL) {
        capt->fptr_push(capt->uptr_push, NULL);
        capt->fptr_push = NULL;
    }
}

static video_capture_t capt = {
    .lck = lock_initial,
    .jcode = {{"libmedia/media", "video_start",  "(IIIIIJ)[I"}, NULL, NULL, 0},
};

void* video_create(enum camera_position_t campos, video_push_t func, void* ctx, video_format* fmt)
{
    errno = EINVAL;
    if (func == NULL) {
        return NULL;
    }

    errno = EEXIST;
    lock(&capt.lck);
    if (capt.fptr_push != NULL) {
        if (capt.run == 0) {
            errno = EAGAIN;
        }
        unlock(&capt.lck);
        return NULL;
    }

    errno = ENOMEM;
    my_handle* handle = handle_attach(&capt, capt_cleanup);
    if (handle == NULL) {
        unlock(&capt.lck);
        return NULL;
    }

    capt.fptr_push = func;
    capt.uptr_push = ctx;
    capt.fmt = fmt;
    capt.seq = 0;
    capt.id = media_id_self;
    capt.campos = campos;
    int64_t* feature = android_feature_address();
    if (campos == back) {
        capt.angle = (intptr_t) feature[back_angle];
        capt.mirror = 0;
    } else {
        capt.angle = (intptr_t) feature[front_angle];
        capt.mirror = 0x80;
    }
    handle_clone(handle);
    capt.self = handle;

    errno = EFAILED;
    if (-1 == video_begin_capt(handle, &capt)) {
        capt.fptr_push = NULL;
        handle_dettach(handle);
        handle = NULL;
    } else {
        capt.run = 1;
        handle_clone(handle);
    }
    unlock(&capt.lck);
    return handle;
}

void video_destroy(void* ptr)
{
    my_handle* handle = (my_handle *) ptr;
    video_capture_t* capt = (video_capture_t *) handle_get(handle);

    if (capt != NULL) {
        lock(&capt->lck);
        capt->run = 0;
        unlock(&capt->lck);
        handle_put(handle);
    }
    handle_dettach(handle);
}

int use_camera(enum camera_position_t campos)
{
    lock(&capt.lck);
    if (capt.run == 0) {
        unlock(&capt.lck);
        return 0;
    }

    my_handle* handle = capt.self;
    handle_clone(handle);
    video_capture_t* captr = (video_capture_t *) handle_get(handle);
    if (captr == NULL) {
        handle_release(handle);
        unlock(&capt.lck);
        return 0;
    }

    if (captr->campos == campos) {
        handle_put(handle);
        handle_release(handle);
        unlock(&capt.lck);
        return 0;
    }
    enum camera_position_t old = captr->campos;

    captr->angle = -1;
    do {
        handle_clone(handle);
        int n = handle_release(handle);
        if (n == 3) {
            break;
        }
        sched_yield();
    } while (1);

    captr->campos = campos;
    int64_t* feature = android_feature_address();
LABEL:
    if (campos == back) {
        captr->angle = (intptr_t) feature[back_angle];
        capt.mirror = 0;
    } else {
        capt.angle = (intptr_t) feature[front_angle];
        capt.mirror = 0x80;
    }

    int n = video_begin_capt(handle, captr);
    if (n == -1) {
        if (old != captr->campos) {
            captr->campos = old;
            goto LABEL;
        }
        capt.run = 0;
        handle_put(handle);
        handle_dettach(handle);
    }
    handle_put(handle);
    unlock(&capt.lck);

    if (captr->campos == old) {
        return -1;
    }
    return n;
}

File: libmedia/media/platform/ios/video.m

#include "v.h"
#include "mbuf.h"
#include "media_buffer.h"
#include "pts.h"
#include "notifier.h"
#include "lock.h"
#include "fmt_in.h"
#include "ios.h"

@implementation video_session
- (void) detachLost:(NSArray *) devices
{
    uint32_t lost[2] = {1, 1};
    for (AVCaptureDevice* device in devices) {
        if ((NSNull *) device == [NSNull null]) {
            continue;
        }

        assert(device != nil && device != NULL);
        NSString* uniq = device.uniqueID;
        assert(uniq != nil && uniq != NULL);

        if ([device position] == AVCaptureDevicePositionBack) {
            if (uniq == camera[0].unique) {
                lost[0] = 0;
            }
        } else {
            if (uniq == camera[1].unique) {
                lost[1] = 0;
            }
        }
    }

    if (lost[0] == 1) {
        if (camera[0].input != NULL) {
            if (current == &camera[0]) {
                [session removeInput:camera[0].input];
                current = NULL;
            }
            [camera[0].input release];
            camera[0].input = NULL;
            camera[0].unique = NULL;
        }
    }

    if (lost[1] == 1) {
        if (camera[1].input != NULL) {
            if (current == &camera[1]) {
                [session removeInput:camera[1].input];
                current = NULL;
            }
            [camera[1].input release];
            camera[1].input = NULL;
            camera[1].unique = NULL;
        }
    }
}

- (void) tryUse:(AVCaptureDevice *) dev camera:(camera_t *) cam uniq:(NSString *) uniq
{
    AVCaptureDeviceInput* input = [AVCaptureDeviceInput deviceInputWithDevice:dev error:nil];
    if (input != nil) {
        [input retain];
        cam->input = input;
        cam->unique = uniq;
    }
}

- (void) addedOrLost:(enum camera_position_t) campos
{
    NSArray* devices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo];
    [self detachLost:devices];

    camera_t* oldcamera = current;
    if (oldcamera != NULL) {
        if (campos == unkown) {
            if (oldcamera == &camera[0]) {
                campos = back;
            } else {
                campos = front;
            }
        }

        if (oldcamera->input == NULL) {
            oldcamera = NULL;
        } else if ((oldcamera == &camera[0] && campos == front) ||
                   (oldcamera == &camera[1] && campos == back))
        {
            [session removeInput:current->input];
            current = NULL;
            oldcamera = NULL;
        }
    }

    for (AVCaptureDevice* device in devices) {
        if ((NSNull *) device == [NSNull null]) {
            continue;
        }

        my_assert(device != nil && device != NULL);
        NSString* uniq = device.uniqueID;
        my_assert(uniq != nil && uniq != NULL);
        AVCaptureDevicePosition pos = device.position;

        if (camera[0].input == NULL && pos == AVCaptureDevicePositionBack) {
            [self tryUse:device camera:&camera[0] uniq:uniq];
        } else if (camera[1].input == NULL && pos == AVCaptureDevicePositionFront) {
            [self tryUse:device camera:&camera[1] uniq:uniq];
        }

        if (camera[0].input != NULL && camera[1].input != NULL) {
            break;
        }
    }

    if (campos == unkown) {
        return;
    }

    if (campos == back && camera[0].input != NULL) {
        current = &camera[0];
    }

    if (current == NULL && camera[1].input != NULL) {
        current = &camera[1];
    }

    if (current == oldcamera) {
        return;
    }
    [session addInput:current->input];
}

- (int) useCamera:(enum camera_position_t) campos
{
    if (campos == back) {
        if (current == &camera[0]) {
            return 0;
        }

        if (camera[1].input == NULL) {
            return -1;
        }
    }

    if (campos == front) {
        if (current == &camera[1]) {
            return 0;
        }

        if (camera[0].input == NULL) {
            return -1;
        }
    }

    if (campos == back) {
        if (current != NULL) {
            [session removeInput:current->input];
        }
        current = &camera[0];
    } else {
        if (current != NULL) {
            [session removeInput:current->input];
        }
        current = &camera[1];
    }
    [session addInput:current->input];
    return 0;
}

- (BOOL) videoSessionInit:(enum camera_position_t) campos
{
    [self addedOrLost:campos];
    if (camera[0].input == NULL && camera[1].input == NULL) {
        return FALSE;
    }

    [session beginConfiguration];
    AVCaptureVideoDataOutput* videoOut = [[AVCaptureVideoDataOutput alloc] init];
    if (videoOut == nil) {
        [session commitConfiguration];
        return FALSE;
    }
    videoOut.alwaysDiscardsLateVideoFrames = YES;

    dispatch_queue_t q = dispatch_queue_create("video", DISPATCH_QUEUE_SERIAL);
    if (q == NULL) {
        [videoOut release];
        [session commitConfiguration];
        return FALSE;
    }

    [videoOut setSampleBufferDelegate:self queue:q];
    dispatch_release(q);

    if ([session canAddOutput:videoOut]) {
        [session addOutput:videoOut];
    } else {
        [videoOut release];
        [session commitConfiguration];
        return FALSE;
    }

    conn = [videoOut connectionWithMediaType:AVMediaTypeVideo];
    if (conn == nil) {
        [session removeOutput:videoOut];
        [videoOut release];
        [session commitConfiguration];
        return FALSE;
    }
    [conn retain];

    [videoOut release];
    [session commitConfiguration];
    return TRUE;
}

static NSString* preset_from_format(camera_t* cam, video_format* fmt, uint32_t* angle)
{
    NSString* preset = NULL;
    if (fmt->pixel->size->width == 480 && fmt->pixel->size->height == 640) {
        preset = AVCaptureSessionPreset640x480;
        *angle = 0;
    } else if (fmt->pixel->size->width == 640 && fmt->pixel->size->height == 480) {
        preset = AVCaptureSessionPreset640x480;
        *angle = 1;
    } else if (fmt->pixel->size->width == 288 && fmt->pixel->size->height == 352) {
        preset = AVCaptureSessionPreset352x288;
        *angle = 0;
    } else if (fmt->pixel->size->width == 352 && fmt->pixel->size->height == 288) {
        preset = AVCaptureSessionPreset352x288;
        *angle = 1;
    } else if (fmt->pixel->size->width == 720 && fmt->pixel->size->height == 1280) {
        preset = AVCaptureSessionPreset1280x720;
        *angle = 0;
    } else if (fmt->pixel->size->width == 1280 && fmt->pixel->size->height == 720) {
        preset = AVCaptureSessionPreset1280x720;
        *angle = 1;
    } else if (fmt->pixel->size->width == 1080 && fmt->pixel->size->height == 1920) {
        preset = AVCaptureSessionPreset1920x1080;
        *angle = 0;
    } else if (fmt->pixel->size->width == 1920 && fmt->pixel->size->height == 1080) {
        preset = AVCaptureSessionPreset1920x1080;
        *angle = 1;
    } else if (fmt->pixel->size->width == 540 && fmt->pixel->size->height == 960) {
        preset = AVCaptureSessionPresetiFrame960x540;
        *angle = 0;
    } else if (fmt->pixel->size->width == 960 && fmt->pixel->size->height == 540) {
        preset = AVCaptureSessionPresetiFrame960x540;
        *angle = 1;
    } else {
        mark("resolution does not support");
    }
    return preset;
}

- (intptr_t) configure:(video_format *)fmt
{
    if (fmt->pixel->csp != csp_nv12ref && fmt->pixel->csp != csp_nv12) {
        errno = EINVAL;
        return -1;
    }

    NSString* preset = preset_from_format(current, fmt, &angle);
    if (preset == NULL) {
        errno = EINVAL;
        return -1;
    }
    reftype = reference_format(fmt);

    NSString* key = (NSString *) kCVPixelBufferPixelFormatTypeKey;
    NSNumber* value = [NSNumber numberWithUnsignedInt: kCVPixelFormatType_420YpCbCr8BiPlanarFullRange];
    if (value == nil) {
        return -1;
    }

    NSDictionary* settings = [NSDictionary dictionaryWithObject:value forKey:key];
    if (settings == nil) {
        return -1;
    }

    [session beginConfiguration];
    AVCaptureVideoDataOutput* output = (AVCaptureVideoDataOutput *) conn.output;
    output.videoSettings = settings;

    if (angle == 0) {
        conn.videoOrientation = AVCaptureVideoOrientationPortrait;
        camera[0].angle[0] = 1;
        camera[0].angle[1] = 0;
        camera[1].angle[0] = 0;
        camera[1].angle[1] = 1;
    } else if (0) {
        camera[0].angle[0] = 2;
        camera[0].angle[1] = 0;
        camera[1].angle[0] = 0;
        camera[1].angle[1] = 0;
        conn.videoOrientation = AVCaptureVideoOrientationLandscapeLeft;
        angle = 0;
    } else {
        camera[0].angle[0] = 0;
        camera[0].angle[1] = 0;
        camera[1].angle[0] = 0;
        camera[1].angle[1] = 2;
        conn.videoOrientation = AVCaptureVideoOrientationLandscapeRight;
        angle = 0;
    }

    // [current->input device].activeVideoMinFrameDuration = CMTimeMake(1, (int32_t) fmt->caparam->fps);
    // [current->input device].activeVideoMaxFrameDuration = CMTimeMake(1, (int32_t) fmt->caparam->fps);
    if (conn.supportsVideoMinFrameDuration) {
        conn.videoMinFrameDuration = CMTimeMake(1, (int32_t) fmt->caparam->fps);
    }

    if (conn.supportsVideoMaxFrameDuration) {
        conn.videoMaxFrameDuration = CMTimeMake(1, (int32_t) fmt->caparam->fps);
    }

    session.sessionPreset = preset;
    [session commitConfiguration];

    format = fmt;
    return 0;
}

- (video_session *) init:(video_push_t) func arg:(void *) ctx pos:(enum camera_position_t) campos
{
    self = [super init];
    if (self == nil) {
        return nil;
    }

    camera[0].unique = NULL;
    camera[0].input = NULL;
    camera[1].unique = NULL;
    camera[1].input = NULL;
    current = NULL;
    session = [[AVCaptureSession alloc] init];
    conn = nil;
    format = NULL;
    if (session == nil) {
        [self dealloc];
        self = nil;
    }

    if (![self videoSessionInit:campos]) {
        [self dealloc];
        self = nil;
    } else {
        uid = media_id_self;
        seq = 0;
        push = func;
        any = ctx;
    }
    return self;
}

- (int32_t) start
{
    if (!session.isRunning) {
        [session startRunning];
    }
    return 0;
}

- (void) stop
{
    if (session.isRunning) {
        [session stopRunning];
    }
}

- (struct my_buffer *) video_buffer_get1:(CVPixelBufferRef) pixelbuf
{
    size_t bytes = CVPixelBufferGetDataSize(pixelbuf);
    struct my_buffer* mbuf = mbuf_alloc_2((uint32_t) (sizeof(media_buffer) + bytes));
    if (mbuf == NULL) {
        return NULL;
    }

    CVReturn ret = CVPixelBufferLockBaseAddress(pixelbuf, kCVPixelBufferLock_ReadOnly);
    if (kCVReturnSuccess != ret) {
        mbuf->mop->free(mbuf);
        return NULL;
    }

    mbuf->ptr[1] = mbuf->ptr[0] + sizeof(media_buffer);
    mbuf->length -= sizeof(media_buffer);
    media_buffer* media = (media_buffer *) mbuf->ptr[0];
    memset(media->vp, 0, sizeof(media->vp));
    media->frametype = 0;
    media->angle = (angle + (uint32_t) current->angle[0]) | current->mirror;
    media->fragment[0] = 0;
    media->fragment[1] = 1;
    media->iden = uid;
    media->seq = (++seq);
    media->pptr_cc = &format->cc;

    bytes = 0;
    size_t nr_panels = CVPixelBufferGetPlaneCount(pixelbuf);

    for (size_t i = 0; i < nr_panels; ++i) {
        media->vp[i].stride = (uint32_t) CVPixelBufferGetBytesPerRowOfPlane(pixelbuf, i);
        media->vp[i].height = (uint32_t) CVPixelBufferGetHeightOfPlane(pixelbuf, i);
        media->vp[i].ptr = mbuf->ptr[1] + bytes;

        uintptr_t length = media->vp[i].stride * media->vp[i].height;
        bytes += length;

        void* addr = CVPixelBufferGetBaseAddressOfPlane(pixelbuf, i);
        memcpy(media->vp[i].ptr, addr, length);
    }

    ret = CVPixelBufferUnlockBaseAddress(pixelbuf, kCVPixelBufferLock_ReadOnly);
    my_assert(kCVReturnSuccess == ret);
    my_assert(mbuf->length >= bytes);
    mbuf->length = bytes;
    return mbuf;
}

- (struct my_buffer *) video_buffer_get2:(CVPixelBufferRef) pixelbuf
{
    struct my_buffer* mbuf = buffer_attach(pixelbuf, 1);
    if (mbuf != NULL) {
        media_buffer* media = (media_buffer *) mbuf->ptr[0];
        media->frametype = 0;
        media->angle = (angle + (uint32_t) (current->angle[0])) | current->mirror;
        media->fragment[0] = 0;
        media->fragment[1] = 1;
        media->iden = uid;
        media->seq = (++seq);
        media->pptr_cc = &format->cc;
    }
    return mbuf;
}

- (void) captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer
                                                               fromConnection:(AVCaptureConnection *)connection
{
    (void) captureOutput;
    (void) connection;

    CVPixelBufferRef pixelbuf = CMSampleBufferGetImageBuffer(sampleBuffer);
    if (pixelbuf == NULL) {
        return;
    }

    struct my_buffer* mbuf = NULL;
    if (reftype == 0) {
        mbuf = [self video_buffer_get1:pixelbuf];
    } else {
        mbuf = [self video_buffer_get2:pixelbuf];
    }

    if (mbuf != NULL) {
        fill_pts(NULL, mbuf);

        video_push_t video_push = NULL;
        uint32_t n = __sync_add_and_fetch(&refs, 1);
        if (n == 2) {
            video_push = push;
        } else {
            my_assert(n == 1);
        }

        if (video_push != NULL) {
            video_push(any, mbuf);
        } else {
            mbuf->mop->free(mbuf);
        }

        n = __sync_sub_and_fetch(&refs, 1);
        if (n == 0 && video_push != NULL) {
            video_push(any, NULL);
        }
    }
}

- (void) dealloc
{
    if (conn != nil) {
        [conn release];
    }

    if (camera[0].input != NULL) {
        [camera[0].input release];
    }

    if (camera[1].input != NULL) {
        [camera[1].input release];
    }

    if (session != nil) {
        [session release];
    }
    [super dealloc];
}

@end

static video_session* vs = NULL;
static lock_t lck = lock_initial;

void* video_create(enum camera_position_t campos, video_push_t func, void* ctx, video_format* fmt)
{
    errno = EEXIST;
    lock(&lck);
    if (vs != NULL) {
        unlock(&lck);
        return NULL;
    }

    errno = ENOMEM;
    video_session* ptr = [[video_session alloc] init:func arg:ctx pos:campos];
    if (ptr == nil) {
        unlock(&lck);
        return NULL;
    }
    vs = ptr;
    vs->camera[0].mirror = 0x00;
    vs->camera[1].mirror = 0x80;

    errno = EFAILED;
    intptr_t n = [vs configure:fmt];
    if (n != -1) {
        notifer_video_register(vs);
        vs->refs = 1;
        [vs start];
    } else {
        [vs release];
        vs = NULL;
    }
    unlock(&lck);

    return vs;
}

void video_destroy(void* ptr)
{
    my_assert(ptr != NULL);
    lock(&lck);
    if (vs != (video_session *) ptr) {
        unlock(&lck);
        return;
    }

    notifer_video_unregister(ptr);
    uint32_t refs = __sync_sub_and_fetch(&vs->refs, 1);

    video_push_t video_push = vs->push;
    void* any = vs->any;

    [vs stop];
    [vs release];
    vs = NULL;
    unlock(&lck);

    if (refs == 0) {
        video_push(any, NULL);
    }
}

int use_camera(enum camera_position_t campos)
{
    lock(&lck);
    if (vs == NULL) {
        unlock(&lck);
        return 0;
    }
    vs->camera[1].angle[0] = vs->camera[1].angle[1];
    int n = [vs useCamera:campos];
    unlock(&lck);
    return n;
}
